valhalla/meili/geometry_helpers.h

```diff
@@ -10,57 +9,37 @@
 #include <valhalla/midgard/constants.h>
 #include <valhalla/midgard/distanceapproximator.h>
 #include <valhalla/midgard/pointll.h>
+#include <valhalla/midgard/util.h>
 
 namespace valhalla {
 namespace meili {
 namespace helpers {
 
 // snapped point, sqaured distance, segment index, offset
-template <typename coord_t>
-std::tuple<coord_t, float, typename std::vector<coord_t>::size_type, float>
-Project(const coord_t& p,
-        const typename std::vector<coord_t>& shape,
-        const midgard::DistanceApproximator& approximator,
+inline std::tuple<midgard::PointLL, float, typename std::vector<midgard::PointLL>::size_type, float>
+Project(const midgard::projector_t& p,
+        const typename std::vector<midgard::PointLL>& shape,
         float snap_distance = 0.f) {
   if (shape.empty()) {
     throw std::invalid_argument("got empty shape");
   }
 
-  coord_t closest_point(shape.front());
-  float closest_distance = approximator.DistanceSquared(closest_point);
+  midgard::PointLL closest_point(shape.front());
+  float closest_distance = p.approx.DistanceSquared(closest_point);
   decltype(shape.size()) closest_segment = 0;
   float closest_partial_length = 0.f;
   float total_length = 0.f;
-  float lon_scale = cosf(p.lat() * midgard::kRadPerDeg);
 
   // for each segment
   for (decltype(shape.size()) i = 0; i < shape.size() - 1; ++i) {
     // project a onto b where b is the origin vector representing this segment
     // and a is the origin vector to the point we are projecting, (a.b/b.b)*b
     const auto& u = shape[i];
     const auto& v = shape[i + 1];
-    auto bx = v.first - u.first;
-    auto by = v.second - u.second;
-    auto bx2 = bx * lon_scale;
-    auto sq = bx2 * bx2 + by * by;
-    const auto scale =
-        sq > 0 ? (((p.first - u.first) * lon_scale * bx2 + (p.second - u.second) * by) / sq) : 0.f;
-    // projects along the ray before u
-    if (scale <= 0.f) {
-      bx = u.first;
-      by = u.second;
-    } // projects along the ray after v
-    else if (scale >= 1.f) {
-      bx = v.first;
-      by = v.second;
-    } // projects along the ray between u and v
-    else {
-      bx = bx * scale + u.first;
-      by = by * scale + u.second;
-    }
```

Comment:
while i was goose-chasing the oddity about map matching test failure i though that maybe it had something to do with the difference in precision between lokis projectiong and meilis. so to remove that concern i refactored them to share the same code.